Issue 3: Code to replicate the zero-touch deployment demo from THINK 2021 presentation (#4)
edgexfoundry-holding/orra: demos/SDO-OH-EXF-dev-simple/README.md

> NOTE: this is based on the [Open Horizon / EdgeX Foundry integration project](https://github.com/edgexfoundry-holding/open-horizon-integration/blob/master/hub/README.md).

* Clone the [ORRA repository](https://github.com/edgexfoundry-holding/orra) and navigate to the SDO-OH-EXF demo folder:
``` shell
git clone https://github.com/edgexfoundry-holding/orra.git
cd orra/demos/SDO-OH-EXF-dev-simple
```
* Publish the service and pattern:
``` shell
hzn exchange service publish -P -f configuration/service.json
hzn exchange pattern publish -f configuration/pattern.json
```
* Create the temporary and configuration folders, copy the EdgeX Foundry service configuration patterns to their expected location:
``` shell
mkdir -p /var/run/edgex/logs
mkdir -p /var/run/edgex/data
mkdir -p /var/run/edgex/consul/data
mkdir -p /var/run/edgex/consul/config
mkdir -p /root/res
chmod -R a+rwx /var/run/edgex
chmod -R a+rwx /root/res
cp res/* /root/res
```
* Publish the deployment policy:
``` shell
hzn exchange deployment addpolicy -f configuration/deployment.policy.json myorg/policy-com.github.joewxboy.horizon.edgex_1.0.1
```

### Step Four: Edit and run the SDO script

> * Go back to your starting folder:

Reviewer: This assumes that the reader has actually installed OH and started from there. However, in case OH was already installed and step 2 is skipped that may very well not be the case (it wasn't for me). I suggest instead to say something like:
```suggestion
* Go to the folder from which you ran the all-in-one Open Horizon installer. If you went through step 2 above, you should be able to get there with:
```

``` shell
cd ../../..
```

> * Edit the `test-sdo.sh` script to change the following lines:

Reviewer: ```suggestion
* Edit the `test-sdo.sh` script that came with the all-in-one Open Horizon install and change the following lines:
```

>     * Approximately at line 203: `hzn voucher import /var/sdo/voucher.json --policy node.policy.json`

Reviewer: Specifying a line number is bound to quickly be wrong, and as a matter of fact it already is. :-) I suggest instead:
```suggestion
    * Towards the end, the line: `hzn voucher import /var/sdo/voucher.json --policy node.policy.json`
```

>     * Change the policy file to `orra/demos/SDO-OH-EXF-dev-simple/configuration/node.policy`

Reviewer: I suggest instead giving the exact line to change it to:
```suggestion
    * To (adjusting the path to the policy file if needed) : `hzn voucher import /var/sdo/voucher.json --policy orra/demos/SDO-OH-EXF-dev-simple/configuration/node.policy`
```

>     * Approximately at line 210: `/usr/sdo/bin/owner-boot-device ibm.helloworld`

Reviewer: ```suggestion
    * Then the line: `/usr/sdo/bin/owner-boot-device ibm.helloworld`
```

>     * Change the pattern from "ibm.helloworld" to `com.github.joewxboy.horizon.edgex`

Reviewer: Can't we have that pattern be part of the ORRA demo repo instead?

Author: Absolutely.  I'd love to do that once EdgeX is OK with what I've done.  As it currently sits, this is a slightly older release of EdgeX.  I'd like to upgrade it to the latest release first.  But your point is well-taken.

* Run the test script:
``` shell
./test-sdo.sh
```
* Make a request to the EdgeX endpoint for the Random Number Virtual Device:
``` shell
watch "curl --silent http://localhost:48080/api/v1/event | jq .[0].readings[0].value"

> ```

Reviewer: I don't think you should use a watch command here. Or you need to change the description to something like "Make requests to the..."

# Demonstration of SDO, OH, and EXF

## Introduction

This is a simple demonstration geared towards developers.  It begins with the Open Horizon all-in-one installation (including SDO services) in a single VM.  Running the modified SDO script, it will boot a simulated device and then install and configure an Open Horizon Agent.  The Agent will then contact the Management Hub and receive a proposed agreement to install and run an EdgeX Foundry (EXF) instance.  EXF, in turn, will be configured to run a Random Number Device service.

## Requirements

> * An x86_64/amd64-based VM with Ubuntu 18.04 Server, 1 vCPU, 4GB RAM, and 20GB disk

Reviewer 2: I have validated these steps on an HP MP9 G4 system with Intel Core i7 processor, 16GB memory, 500 GB storage, and Ubuntu 18.04.5 desktop.

Author: Good point.  I've added that info and left room for others to add to the list of systems that they've confirmed this on.

## Instructions

### Step One: Prepare the OS

* Boot the VM and log in.  Become root with `sudo -i`.
* **Optional** - Update the OS and install any pre-requisites
``` shell
apt-get -y update
apt-get -y upgrade
apt-get -y install jq
```

### Step Two: Install Open Horizon

> If Open Horizon is already installed and configured properly, skip step two.

> 

Reviewer 2: This step will install docker and docker-compose if those are not present on the system.

Author: Thanks.  I've added that note.

* **Optional** - Determine the VM's local IP address and `export HZN_LISTEN_IP=<local-IP-address-here>` if the network connection is bridged.  If the IP address is accessible outside of the VM, then you will be able to connect to any exposed ports from outside the VM.  Without this step, the services will only be available inside the VM.
* Install the all-in-one Open Horizon instance by following [these instructions](https://github.com/open-horizon/devops/tree/master/mgmt-hub).
* **NOTE** Save the credentials shown at the end of the installation process.  You will need these later.
* Set the following environment variables to connect to the Exchange and to allow publishing.  Replace the `<values>` with the actual value indicated.  Also add them to your shell's RC file if you'd like them to persist beyond this session.
``` shell
export HZN_ORG_ID=myorg
export HZN_EXCHANGE_USER_AUTH=admin:<insert value from installation credentials>
export HZN_DEVICE_TOKEN=<insert value from installation credentials or use random string>
export HZN_DEVICE_ID=node1

> export HZN_EXCHANGE_URL=http://<local-IP-address-here>:3090/v1/

Reviewer 2: The "local-IP-address-here" needs to be "127.0.0.1" for this to work.  Putting in the actual local IP address of my system does not work.

Reviewer: Same here.

Author: Sounds like this was confusing.  I've added wording that you should use the loopback address if you did not expose your internal network address with a bridged network adapter in a VM.

```

### Step Three: Configure and publish the definition files for the EXF Service

* Clone the EdgeX Foundry integration repository and navigate to the `configs` folder:
``` shell
git clone https://github.com/edgexfoundry-holding/orra.git
cd orra/demos/SDO-OH-EXF-dev-simple
```
* Publish the service and pattern:
``` shell
hzn exchange service publish -P -f configuration/service.json
hzn exchange pattern publish -f configuration/pattern.json
```
* Create the temporary and configuration folders, copy the EdgeX Foundry service configuration patterns to their expected location:
``` shell
mkdir -p /var/run/edgex/logs
mkdir -p /var/run/edgex/data
mkdir -p /var/run/edgex/consul/data
mkdir -p /var/run/edgex/consul/config
mkdir -p /root/res
chmod -R a+rwx /var/run/edgex
chmod -R a+rwx /root/res
cp res/* /root/res
```
* Publish the deployment policy:
``` shell
hzn exchange deployment addpolicy -f configuration/deployment.policy.json myorg/policy-com.github.joewxboy.horizon.edgex_1.0.1
```

### Step Four: Edit and run the SDO script

* Go back to your starting folder:
``` shell
cd ../../..
```
* Edit the `test-sdo.sh` script to change the following lines:

>     * Approximately at line 203: `hzn voucher import /var/sdo/voucher.json --policy node.policy.json`

Reviewer 2: Should find the right place earlier in this flow to indicate when was the "test-sdo.sh" script got created.  Was it during open horizon install?

Reviewer: I agree, I suggested making it clear this is found within the all-in-one OH install.

Author: :) I've added that information in a note.  Thanks.

    * Change the policy file to `orra/demos/SDO-OH-EXF-dev-simple/configuration/node.policy`
    * Approximately at line 210: `/usr/sdo/bin/owner-boot-device ibm.helloworld`
    * Change the pattern from "ibm.helloworld" to `com.github.joewxboy.horizon.edgex`
* Run the test script:
``` shell
./test-sdo.sh

> ```

Reviewer 2: Suggest to add some commentary during run of this script as to what is happening.

Reviewer: I agree and it'd be good to let people know not to worry if they see:
```
26-04-2021 12:50:39.072 WARN  [main] org.sdo.pri.device.DeviceApp.sslContext - UNSAFE: no-op TrustManager installed
```
and
```
26-04-2021 12:50:53.800 ERROR [main] o.s.p.DeviceTransferOwnershipClient.call - (POST http://127.0.0.1:8040/mp/113/msg/30) 500 {"timestamp":"2021-04-26T12:50:53.370+00:00","status":500,"error":"Internal Server Error","message":"","path":"/mp/113/msg/30"}
26-04-2021 12:50:53.879 INFO  [main] o.s.p.DeviceTransferOwnershipClient.call - All rendezvous instructions exhausted. Pausing until Apr 26, 2021, 12:52:48 PM
```
It eventually worked out all right after that pause but it left me wondering what had gone wrong while it was on pause. It would actually be good to tell people what to look for for a sign of successful run.

Author: Good point.  Added wording to that effect.
